src/Data/Conduit/Tar.hs

```diff
@@ -270,3 +245,482 @@ withEntries :: MonadThrow m
+    hNamePrefix <- encodeShort 155 headerFileNamePrefix
+    return
+        ( toLazyByteString $
+          hNameSuffix <>
+          hFileMode <> word8 0 <>
+          hOwnerId <> word8 0 <>
+          hGroupId <> word8 0 <>
+          hPayloadSize <> word8 0 <>
+          hTime <> word8 0
+        , toLazyByteString $
+          word8 headerLinkIndicator <>
+          hLinkName <>
+          hMagicVersion <>
+          hOwnerName <>
+          hGroupName <>
+          hDevMajor <> word8 0 <>
+          hDevMinor <> word8 0 <>
+          hNamePrefix <>
+          byteString (S.replicate 12 0)
+        )
+  where
+    encodeDevice 0     = return $ byteString $ S.replicate 7 0
+    encodeDevice devid = encodeOctal 7 devid
+
+
+-- | Encode a `ShortByteString` with an exact length, NUL terminating if it is
+-- shorter, but throwing `TarCreationError` if it is longer.
+encodeShort :: MonadThrow m => Int -> ShortByteString -> m Builder
+encodeShort !len !sbs
+    | lenShort <= len = return $ byteString $ fst $ S.unfoldrN len maybeShort 0
+    | otherwise =
+        throwM $
+        TarCreationError $ "Can't fit '" ++ S8.unpack (fromShort sbs) ++ "' into the tar header"
+  where
+    lenShort = SS.length sbs
+    maybeShort !i
+        | i < lenShort = Just (SS.index sbs i, i + 1)
+        | otherwise = Just (0, i + 1)
+
+
+-- | Encode a number in 8base padded with zeros. Throws `TarCreationError` when overflows.
+encodeOctal :: (Show a, Integral a, MonadThrow m) => Int -> a -> m Builder
+encodeOctal !len !val = do
+    enc <- case S.unfoldrN len toOctal val of
+        (valStr, Just 0) -> return valStr
+        over -> throwM $ TarCreationError $ "<encodeOctal>: Tar value overflow: " ++ show over
+    return (byteString $ S.reverse enc)
+  where
+    toOctal x =
+        let !(q, r) = x `quotRem` 8
+        in Just (fromIntegral r + 48, q)
+
+
+
+-- | Produce a ByteString chunk with NUL characters of the size needed to get up
+-- to the next 512 byte mark in respect to the supplied offset and return that
+-- offset incremented to that mark.
+yieldNulPadding :: Monad m => FileOffset -> ConduitM i ByteString m FileOffset
+yieldNulPadding n = do
+    let pad = blockSize - (n `mod` blockSize)
+    if pad /= blockSize
+        then yield (S.replicate (fromIntegral pad) 0) >> return (n + pad)
+        else return n
+
+
+
+
+-- | Handle tar payload, while validating its size and padding it to the full
+-- block at the end.
+tarPayload :: MonadThrow m =>
+              FileOffset -- ^ Received payload size
+           -> Header -- ^ Header for the file that we are currently recieving the payload for
+           -> (FileOffset -> ConduitM (Either a ByteString) ByteString m FileOffset)
+           -- ^ Continuation for after all payload has been received
+           -> ConduitM (Either a ByteString) ByteString m FileOffset
+tarPayload size header cont
+    | size == headerPayloadSize header = cont (headerOffset header + blockSize)
```

Comment:
This would be another good location for worker/wrapper transform, since it appears that `header` and `cont` remain unchanged and only `size` is being updated each step thought.

Reply:
You'd expect GHC would be able optimize such a common pattern. I made the change here as well